File: server/utils/event_bus.go
package utils

import (
	"sync"
)

// Event represents an event with a name and context.
type Event struct {
	Name    string
	Context map[string]interface{}
}

// EventListener represents an event listener.
type EventListener interface {
	HandleEvent(Event)
}

var eventMutex = sync.RWMutex{}
var eventListeners = make(map[string][]EventListener)

// RegisterListener registers a new listener to the event bus.
func RegisterListener(eventName string, listener EventListener) {
	eventMutex.Lock()
	defer eventMutex.Unlock()

	if _, exists := eventListeners[eventName]; !exists {
		eventListeners[eventName] = make([]EventListener, 0)
	}

	eventListeners[eventName] = append(eventListeners[eventName], listener)
}

// PublishEvent publishes an event with a given context to all registered listeners.
func PublishEvent(event Event) {
	eventMutex.RLock()
	defer eventMutex.RUnlock()

	if _, exists := eventListeners[event.Name]; !exists {
		return
	}

	for _, listener := range eventListeners[event.Name] {
		// Each listener runs in its own goroutine to prevent blocking.
		go listener.HandleEvent(event)
	}
}

File: server/utils/event_bus.py
import threading

class Event:
    def __init__(self, name, context):
        self.name = name
        self.context = context

class EventListener:
    def handle_event(self, event: Event):
        pass

event_mutex = threading.Lock()
event_listeners = {}

def register_listener(event_names, listener: EventListener):
    with event_mutex:
        for name in event_names:
            if name not in event_listeners:
                event_listeners[name] = []
            event_listeners[name].append(listener)

def publish_event(event: Event):
    with event_mutex:
        listeners = event_listeners.get(event.name, [])[:]
    for listener in listeners:
        threading.Thread(target=listener.handle_event, args=(event,)).start()
